safe-transaction-client/src/lib.rs

```diff
@@ -0,0 +1,343 @@
+                // DOMAIN_SEPARATOR_TYPEHASH
+                U256::from_str(
+                    "0x47e79534a245952e8b16893a336b85a3d9ea9fa8c573f3d803afb92a79469218"
+                )
+                .unwrap(),
+                U256::from(signer.chain_id()),
+                self.safe_address,
+            ])));
+            // safeTxHash
+            hasher.update(&keccak256(&ethers::abi::encode(tokenize![
+                // SAFE_TX_TYPEHASH
+                U256::from_str(
+                    "0xbb8310d486368db6bd6f849402fdd73ad53d316b5a4b2644ad6efe0f941286d8"
+                )
+                .unwrap(),
+                self.to,
+                self.value,
+                keccak256(&self.data),
+                self.operation as u8,
+                U256::zero(),    // safe_tx_gas
+                U256::zero(),    // base_gas
+                U256::zero(),    // gas_price
+                Address::zero(), // gas_token
+                Address::zero(), // refund_receiver
+                self.nonce,
+            ])));
+            hasher.finalize(&mut output);
+            output
+        });
+
+        let SignedSafeTxHash { signature, .. } = sign_tx_hash(signer, safe_tx_hash).await?;
+
+        Ok(SignedSafeTx {
+            inner: self,
+            sender: signer.address(),
+            safe_tx_hash,
+            signature,
+        })
+    }
+}
+
+pub struct SignedSafeTxHash {
+    safe_tx_hash: TxHash,
+    signature: Signature,
+}
+
+pub async fn sign_tx_hash<S>(
+    signer: &S,
+    safe_tx_hash: TxHash,
+) -> Result<SignedSafeTxHash, Error<S::Error>>
+where
+    S: Signer,
+    Error<S::Error>: From<S::Error>,
+{
+    let mut signature = signer.sign_message(safe_tx_hash).await?;
+
+    // signature.recover expects signature.v ∈ {27, 28}
+    if signature.v < 2 {
+        signature.v += 27;
+    }
+    assert!(
+        [27, 28].contains(&signature.v),
+        "BUG: `Signer::sign_message` produced an invalid `v` field."
+    );
+    assert_eq!(
+        Some(signer.address()),
+        signature.recover(&safe_tx_hash[..]).ok(),
+        "BUG: `Signature::recover` failed to recover our signing \
+            address... did `Signer::sign_message` fail to add the prefix?"
+    );
+    signature.v += 4;
+
+    Ok(SignedSafeTxHash {
+        safe_tx_hash,
+        signature,
+    })
+}
+
+#[test]
+#[cfg(feature = "run-test-against-api")]
+fn check() {
+    futures_executor::block_on(async {
+        use rand_chacha::rand_core::SeedableRng;
+
+        // super secure RNG for tests
+        let mut fake_rng =
+            rand_chacha::ChaCha20Rng::from_seed(*b"0123456789abcdef0123456789abcdef");
+        let wallet = LocalWallet::new(&mut fake_rng).with_chain_id(4u64);
+
+        // [x] propose a new transaction
+
+        let value = U256::from(0u64);
+        let _data_unanchor =
+            hex::decode("cbd566010000000000000000000000000000000000000000000000000000000000000000")
+                .unwrap();
```

Comment:
Unused var?

Reply:
Easy to swap between the unanchor and anchor data examples. It is effectively a type-checked comment. I've restructured it for clarity of the choices.